### DISCUSSION TITLE
Retrieving Current Temps via API

### DISCUSSION THREAD

**Discussion by zigsphere** [weather-gov/api#558] [Q&A]
Hello all,

I've been playing around with the API and reading the documentation. My understanding is that the API can retrieve the forecast, but what can be used to retrieve the current temperature for the lat/long specified? The forecast is usually for a future time/day.

**Comment by dwhitemv25** [comment]
Do you want a measured temperature, or a predicted temperature from a model?

**Reply by zigsphere** [reply]
I dont want predicted temp. I just want current temp for a particular area at the time the API is called. Is that possible with this API?

E.g - someone on my site visits a page for a particular city. Site pulls current temp for that city with the API.

**Reply by dwhitemv25** [reply]
It takes a little more footwork with the weather.gov API to answer that kind of question, especially since no geocoding features are provided. This is how you'd get a measured temperature near to a given point.

If you call `/points/XXXXXX,YYYYYYY`, with the lat/lon, the `properties/observationStations` key provides another API call, `/gridpoints/AA,BB/stations`, that returns a list of weather stations near that point, sorted by distance. You can take a station from that list, call `/stations/STATIONID/observations`, and that gives you the weather at that station. 

In an urban area its more likely to find suitable nearby stations versus a rural area where stations are few and far between.

**Reply by zigsphere** [reply]
Thank you!

**Reply by DavidBerdik** [reply]
@dwhitemv25 When you state "predicted temperature from a model", are you referring to a reported current temperature for scenarios in which there is no station nearby? If so, how would I go about obtaining that through the API?

**Reply by dwhitemv25** [reply]
To answer the question bluntly ... you don't. The NWS API does not provide that kind of data. (At least as of the date of this post.)

By definition, there is no "reported current temperature" where "there is no station nearby." If a website is giving you a "current temperature" at your location and you don't have a reporting station there, one of 2 things is happening:

* The site is producing a synthetic value by interpolating between reporting stations; or
* The site is referencing a forecast model and giving you the temperature the model thinks it should be at or near your location.

In the case of the NWS API, it only gives you current temperatures for reporting stations the NWS tracks, which are usually major installations like airports. You can query the calls I mentioned in my above post to get the list of those reporting stations. Hopefully one is relatively near (spatially and climatologically) to the point you're interested in.

If you want:
* An interpolated value: You'll have to do that math yourself.
* A model value: You'll have to download the model data files and pull from those.

... or find another service that will do that for you.

**Reply by DavidBerdik** [reply]
Understood. Thank you for the thorough explanation. Unfortunately, for my use case, it seems like it would best to find a different API to use.

[Open-Meteo](https://open-meteo.com/) seems like a good alternative.

**Reply by scadergit** [reply]
This is no different than all weather APIs. Current weather has to derive from an observation station. The data endpoints on the forecast application were limited to MADIS (major observation stations), but the API contains the same you'd find elsewhere.

**Reply by illuminated-script** [reply]
> It takes a little more footwork with the weather.gov API to answer that kind of question, especially since no geocoding features are provided. This is how you'd get a measured temperature near to a given point.
> 
> If you call `/points/XXXXXX,YYYYYYY`, with the lat/lon, the `properties/observationStations` key provides another API call, `/gridpoints/AA,BB/stations`, that returns a list of weather stations near that point, sorted by distance. You can take a station from that list, call `/stations/STATIONID/observations`, and that gives you the weather at that station.
> 
> In an urban area its more likely to find suitable nearby stations versus a rural area where stations are few and far between.

This is exactly what I was looking for with the project I'm working on. Thank you!

**Comment by DavidBerdik** [comment]
@scadergit I understand that all data is derived from observation stations, but for my use case, I need an API that does the automation of producing the synthetic values for areas which don't have lots of observation stations. @dwhitemv25 has explained that the NWS API doesn't do that. It looks like Open-Meteo does.

**Reply by scadergit** [reply]
I would suggest Open-Metro does the same as us, we just make it more transparent. For example, if you go to weather underground and choose a location, the current temp is associated to the nearest station. That's common practice, but I haven't delved into the Open-Metro code to verify.

**Reply by scadergit** [reply]
[I asked open-metro how they calculate observtions](https://github.com/open-meteo/open-meteo/discussions/161), and I stand corrected. They took the novel approach to simply use computer models _for everything_. So no data of theirs is actual weather, but that would simplify the approach to current weather, which is based on actual weather.

**Reply by DavidBerdik** [reply]
Right, and if I correctly understand what was said earlier, I can technically do the same thing using the NWS API, but I would have to do all the computations for it myself, which is not feasible for my use case.

**Reply by scadergit** [reply]
The API doesn't provide data models. It only provides actual weather from observation stations and forecast weather from offices. Model data is provided by https://mag.ncep.noaa.gov/, and as open-metro pointed out, parsing that is not always simple.

**Reply by DavidBerdik** [reply]
Yep! For my use case, I've decided to use Open-Meteo since it handles the heavy lifting with that, but I still use the NWS API for handling severe weather alerts.
